### PR TITLE
[ENH, MRG] Add mne.viz.Brain.add_volume_labels

### LINKED ISSUE #9520: [ENH] Add "roi" feature to `mne.viz.Brain`
One thing that is nice for plotting in 3D with stereoEEG is to be able to plot ROIs. There is already a `surfaces` keyword argument in `mne.viz.plot_alignment` does something similar. We've just added the marching cubes implementation in #9484 so it would be as simple as passing an atlas in the subjects directory, locating the aseg file based on the atlas, running marching cubes which only takes a second at most and then plotting the surface.

A call could look like this:

```
mne.viz.plot_alignment(raw.info, atlas='desikan-killiany', rois=('hippocampus', 'putamen', 'caudate'))
```

### PR DESCRIPTION
**Pull request by alexrockhill**
Closes #9520

Here is a very, very rough draft of what I was thinking, I didn't want to go too far in a bad direction so if you have a second to look at it @larsoner, that would be very much appreciated.

Basically, the ROIs are a very different kind of surfaces than the SUBJECTS_DIR/SUBJECT/surf surfaces and so I think need their own class but I think inheritance works nicely; they are named very differently and don't all have hemispheres.

### CODE REVIEW

File: mne/source_space.py
Comment (larsoner) [inline mne/source_space.py:46]: Try to avoid these big namespace moves in PRs (unless they're dedicated just to that), it makes the diff/API discussion harder to review and discuss

File: mne/surface.py
Comment (larsoner) [inline mne/surface.py:1708]: As discussed on Discord, I don't think we want/need curvatures for volume ROIs as they likely aren't as meaningful as they are for cortex (where the sulcal/gyral distinction is important)

File: mne/surface.py
Comment (larsoner) [inline mne/surface.py:1758]: aseg should probably default to `aparc+aseg`

Also worth seeing how much code can be refactored from `find_pos_in_annot(pos, subject='fsaverage', annot='aparc+aseg'` in `label.py`

File: mne/viz/_brain/_brain.py
Comment (larsoner) [inline mne/viz/_brain/_brain.py:262]: I would not add new arguments to the constructor. Instead, a new method `brain.add_aseg(...)` (to mirror `brain.add_parcellation`) would be good, e.g.:
```
def add_aseg(aseg='aseg.mgz', labels=None)
```
and by default `labels=None` means "use all labels in the given parcellation"

File: mne/viz/_brain/surface.py
Comment (larsoner) [inline mne/viz/_brain/surface.py:130]: I wouldn't add anything new to `_Surface` -- this is really for brain left/right hemi surfaces (feel free to update the docs to make this clear). For any other surfaces (including the marching-cube-created surfaces) they can be plotted directly using the renderer mesh methods.

File: mne/viz/_brain/surface.py
Comment (larsoner) [inline mne/viz/_brain/surface.py:179]: I wouldn't subclass surface but rather create a new class. I don't think they'll really need to share many properties.

For now I would just implement the marching cubes + rendering directly in _Brain (maybe in a helper function that takes a volume and ROI values and maps these to visible VTK objects), I don't think a class will really help much here

File: mne/surface.py
Comment (larsoner) [inline mne/surface.py:1754]: Possibly unrelated cruft since it's not used in the example? Probably makes sense to split into multiple PRs:

1. one that makes it easy to do this electrode mapping
2. another that deals with extracting ROIs associated with electrodes
3. another that deals with visualization of volumetric atlas / aseg in _Brain

File: mne/viz/_brain/_brain.py
Comment (alexrockhill) [reply mne/viz/_brain/_brain.py:262]: Great thanks, that's why I asked now :)

**Comment by alexrockhill** [comment]
Ok this is missing all the brain stuff but I was a bit too tired to figure it out tonight. Looks like it works as a minimal version though. I'll add some tests tomorrow or maybe I could get a little help finishing this one because the `mne.viz.Brain` class is quite complicated and I'm not sure how much I really need to dive in for something this simple. Maybe we can talk about it tomorrow. I'll go ahead and start on the GUI tomorrow too.

Screen Shot 2021-07-14 at 11 05 46 PM

**Comment by larsoner** [comment]
Using:
```
import mne
subjects_dir = mne.datasets.sample.data_path() + '/subjects'
brains = list()
brain = mne.viz.Brain(
    'fsaverage', 'both', 'pial',
    alpha=0.1, cortex='low_contrast', subjects_dir=subjects_dir)
brain.show_view(dict(azimuth=90, elevation=90))
brain.add_aseg()
brain.enable_depth_peeling()
```
I get:

Screenshot from 2021-07-15 14-37-20

So I added some smoothing and now I get:

Screenshot from 2021-07-15 15-08-55

I'll push, then @alexrockhill I think if you add this to `tutorials/inverse/60_visualize_stc.py` to show that we can now `add_aseg` in addition to `add_annotation` then I think we're close to done here!

**Comment by alexrockhill** [comment]
Awesome! Looks great!!

Sure, I can do the last bit when you push your version.

**Comment by larsoner** [comment]
I also feel bad for the poor neglected cerebellum so I added it to the defaults:

Screenshot from 2021-07-15 15-27-53

**Comment by larsoner** [comment]
I'm done, feel free to push!

**Comment by alexrockhill** [comment]
Wow you put labels on each of the default lut indices, thanks!

**Comment by larsoner** [comment]
Yeah it was a quick copy-paste-trim from the FreeSurfer LUT file. vscode makes it really easy to edit N lines in an identical way :)

**Comment by alexrockhill** [comment]
Hmmm, some camera issues I think

brain2

**Comment by alexrockhill** [comment]
Screen Shot 2021-07-15 at 3 53 55 PM

Hmmm the electrodes seem to be out-of-scale as well... it might be easier to mock up a quick `add_sensors` that doesn't support MEG but I don't want to make problems/get in the way of @GuillaumeFavelier

**Comment by larsoner** [comment]
You need units='m' in the brain constructor

**Comment by larsoner** [comment]
(plot_alignment only ever uses meters but brain defaults to millimeters for consistency with freesurfer and pysurfer)

**Comment by larsoner** [comment]
Last run CircleCI complained:
```
/home/circleci/project/mne/viz/_brain/_brain.py:docstring of mne.viz._brain._brain.Brain.add_aseg:13: WARNING: py:func reference target not found: mne.get_montage_rois
/home/circleci/project/mne/viz/_brain/_brain.py:docstring of mne.viz._brain._brain.Brain.add_aseg:13: WARNING: undefined label: freesurfer lookup table
/home/circleci/project/mne/viz/_brain/_brain.py:docstring of mne.viz._brain._brain.Brain.add_aseg:32: WARNING: py:func reference target not found: pyvista.BasePlotter.add_legend
changes/latest.inc:102: WARNING: py:func reference target not found: mne.viz.Brain.add_eeg_sensors
```
Not sure if you fixed those in the latest rebase or not...

**Comment by alexrockhill** [comment]
> Last run CircleCI complained:
> 
> ```
> /home/circleci/project/mne/viz/_brain/_brain.py:docstring of mne.viz._brain._brain.Brain.add_aseg:13: WARNING: py:func reference target not found: mne.get_montage_rois
> /home/circleci/project/mne/viz/_brain/_brain.py:docstring of mne.viz._brain._brain.Brain.add_aseg:13: WARNING: undefined label: freesurfer lookup table
> /home/circleci/project/mne/viz/_brain/_brain.py:docstring of mne.viz._brain._brain.Brain.add_aseg:32: WARNING: py:func reference target not found: pyvista.BasePlotter.add_legend
> changes/latest.inc:102: WARNING: py:func reference target not found: mne.viz.Brain.add_eeg_sensors
> ```
> 
> Not sure if you fixed those in the latest rebase or not...

That's going to depend on #9545 to merge first but other than that, yes taken care of

File: mne/viz/_brain/tests/test_brain.py
Comment (larsoner) [inline mne/viz/_brain/tests/test_brain.py:316]: to speed this up can you set `rois` to some small list of one or two? This will also have a nice side-effect of hitting the `legend` lines

File: mne/viz/_brain/_brain.py
Comment (larsoner) [inline mne/viz/_brain/_brain.py:391]: I don't think these should default to None since most of the time people *do* want surfaces. Probably `hemi='both', surfaces='pial'` is the most sensible default?

File: mne/viz/_brain/_brain.py
Comment (larsoner) [inline mne/viz/_brain/_brain.py:2347]: Can be more compact as:
```suggestion
        legend : bool | None | dict
            Add a legend displaying the names of the ``rois``. Default (None)
            is ``True`` if the number of ``rois`` is 10 or fewer. Can also be a
            dict of `kwargs` to pass to :meth:`pyvista.BasePlotter.add_legend`.
```

File: mne/viz/_brain/_brain.py
Comment (larsoner) [inline mne/viz/_brain/_brain.py:2379]:
```suggestion
        # convert legend to False or dict of kwargs to pass
        _validate_type(legend, (None, bool, dict), 'legend')
        if legend is None:
            legend = len(rois) < 11
        if not isinstance(legend, dict):
            legend = dict() if legend else False
```

File: mne/viz/_brain/_brain.py
Comment (larsoner) [inline mne/viz/_brain/_brain.py:2400]:
```suggestion
        if legend is not False:
            self._renderer.plotter.add_legend(
                list(zip(rois, colors)), **legend)
```

Comment (alexrockhill) [reply mne/viz/_brain/_brain.py:2400]: It can't be `None`, it would have gotten reassigned earlier so I think zen is not to do this.

Comment (larsoner) [reply mne/viz/_brain/_brain.py:2400]: It can't be None but it can be `dict()` and:
```
>>> bool(dict()) is False
True
```

Comment (alexrockhill) [reply mne/viz/_brain/_brain.py:2400]: Oh sorry I was scrolling up from the bottom and I didn't see your refactoring comment above

**Comment by larsoner** [comment]
@alexrockhill this needs a rebase or merge. I went to do it but I think you'll have ideas for how best to combine `add_aseg` with the other tutorial changes.

Also based on naming consistency I suppose this should be `add_volume_labels`

**Comment by larsoner** [comment]
@alexrockhill I pushed a commit that cleans up the smoothing and makes the marching cubes faster for labeled volumes following the VTK example https://kitware.github.io/vtk-examples/site/Cxx/Medical/GenerateModelsFromLabels/

**Comment by larsoner** [comment]
... so I think all we need here is a rebase or merge with upstream/main, and rename `add_aseg` to `add_volume_labels` then we should be good!

**Comment by alexrockhill** [comment]
> @alexrockhill I pushed a commit that cleans up the smoothing and makes the marching cubes faster for labeled volumes following the VTK example https://kitware.github.io/vtk-examples/site/Cxx/Medical/GenerateModelsFromLabels/

I don't see this commit, did it make it in? Sorry if I forced a push over it, hopefully you still have a local copy

**Comment by larsoner** [comment]
> I don't see this commit, did it make it in? Sorry if I forced a push over it, hopefully you still have a local copy

No it didn't, but I just pushed that (and some minor fixes to it) plus a change to the `10_` example so that CircleCI can run it and we can verify it fixes the failure

File: mne/surface.py
Comment (larsoner) [inline mne/surface.py:1674]: This extra text should go in `docs.py` not here (it's the same in both cases)

File: mne/utils/docs.py
Comment (larsoner) [inline mne/utils/docs.py:1695]:
```suggestion
smooth : float in [0, 1)
```

File: mne/viz/_3d.py
Comment (larsoner) [inline mne/viz/_3d.py:2458]: Can either revert this and related `Brain(..., hemi=None)` changes, or we need to add a test that uses `hemi=None` in `test_brain.py`

**Review by larsoner — CHANGES_REQUESTED** [review]

File: mne/surface.py
Comment (alexrockhill) [reply mne/surface.py:1674]: Great, will do

File: mne/viz/_3d.py
Comment (alexrockhill) [reply mne/viz/_3d.py:2458]: Ok, a test should be easy

File: mne/viz/_brain/_brain.py
Comment (larsoner) [inline mne/viz/_brain/_brain.py:2346]:
```suggestion
            Can also be a dict of ``kwargs`` to pass to
```

**Review by larsoner — COMMENTED** [review]
This is probably going to make Sphinx angry

**Comment by alexrockhill** [comment]
I'm still getting

```
/Users/alexrockhill/projects/mne-python/mne/viz/_brain/_brain.py:docstring of mne.viz._brain._brain.Brain.add_volume_labels:13: WARNING: undefined label: freesurfer lookup table
```

Even though it's in `doc/links.inc`.

This is for `:ref:`freesurfer lookup table`. Do you know how to fix this?

**Comment by larsoner** [comment]
`links.inc` must be included manually using a `.. include::` directive. We can add it to the autosummary template I think (?). I'll give it a shot and push if it works. But actually another option would be to add a glossary entry, which we should maybe do anyway. I'll just go with that option...

**Review by larsoner — APPROVED** [review]
LGTM, let's merge once the CIs come back happy!

**Comment by larsoner** [comment]
@alexrockhill is this expected?
```
2021-07-19 14:36:50.010 (  56.876s) [        4049D740]vtkWindowedSincPolyData:1426   ERR| vtkWindowedSincPolyDataFilter (0x8d78720): No data to smooth!
WARNING: /home/larsoner/python/mne-python/tutorials/clinical/20_seeg.py failed to execute correctly: Traceback (most recent call last):
  File "/home/larsoner/python/mne-python/tutorials/clinical/20_seeg.py", line 178, in <module>
    brain.add_volume_labels(aseg='aparc+aseg', labels=labels)
  File "/home/larsoner/python/mne-python/mne/viz/_brain/_brain.py", line 2391, in add_volume_labels
    warn(f'Value not found for label {repr(label)} in: '
  File "/home/larsoner/python/mne-python/mne/utils/_logging.py", line 358, in warn
    warnings.warn_explicit(
RuntimeWarning: Value not found for label 'ctx-lh-insula' in: /home/larsoner/mne_data/MNE-sample-data/subjects/fsaverage/mri/aparc+aseg.mgz
```

**Comment by alexrockhill** [comment]
> @alexrockhill is this expected?
> 
> ```
> 2021-07-19 14:36:50.010 (  56.876s) [        4049D740]vtkWindowedSincPolyData:1426   ERR| vtkWindowedSincPolyDataFilter (0x8d78720): No data to smooth!
> WARNING: /home/larsoner/python/mne-python/tutorials/clinical/20_seeg.py failed to execute correctly: Traceback (most recent call last):
>   File "/home/larsoner/python/mne-python/tutorials/clinical/20_seeg.py", line 178, in <module>
>     brain.add_volume_labels(aseg='aparc+aseg', labels=labels)
>   File "/home/larsoner/python/mne-python/mne/viz/_brain/_brain.py", line 2391, in add_volume_labels
>     warn(f'Value not found for label {repr(label)} in: '
>   File "/home/larsoner/python/mne-python/mne/utils/_logging.py", line 358, in warn
>     warnings.warn_explicit(
> RuntimeWarning: Value not found for label 'ctx-lh-insula' in: /home/larsoner/mne_data/MNE-sample-data/subjects/fsaverage/mri/aparc+aseg.mgz
> ```

No, but all the examples look just fine, I think maybe without smoothing it's too small, I just deleted it as a testing label a commit or two ago

**Comment by alexrockhill** [comment]
The seeg example could use the focal point reset. Is there any way to do that via the brain object without `brain._renderer.plotter.camera.focal_point = ...`?

**Comment by alexrockhill** [comment]
And it would be nice to color the electrodes differently but that's only available in https://github.com/mne-tools/mne-python/pull/9585 not `mne.viz.plot_alignment`.

On another day though...

**Comment by larsoner** [comment]
>  I just deleted it as a testing label a commit or two ago

Can you update the example, too? You can see in the traceback the error comes when running the example (CircleCI will soon complain the same way...)

FYI it's good practice locally when iterating to do things like `PATTERN=clinical make html_dev-pattern-memory`, which is how I found it. That way you don't have to wait for CIs, especially when their backlog is long like right now

> brain._renderer.plotter.camera.focal_point = 

`set_3d_view` has a `focalpoint` option

https://mne.tools/dev/generated/mne.viz.set_3d_view.html

But maybe we're missing something in the viz calls if the view is being messed up / changed, probably a reset_camera=False somewhere...

**Comment by larsoner** [comment]
... just realized you asked about `Brain`. We should expose `focalpoint` in https://mne.tools/stable/generated/mne.viz.Brain.html#mne.viz.Brain.show_view

**Comment by larsoner** [comment]
> And it would be nice to color the electrodes differently but that's only available in #9585 not mne.viz.plot_alignment.

In this case I would favor a PR to `plot_alignment` first that allows channel-by-channel color choosing first, as we'll want it there, too. Then later refactoring as we've talked about.

**Comment by alexrockhill** [comment]
> ... just realized you asked about `Brain`. We should expose `focalpoint` in https://mne.tools/stable/generated/mne.viz.Brain.html#mne.viz.Brain.show_view

Yeah I can open an issue for that too, I was able to just change the view angles for now.

**Comment by larsoner** [comment]
> Yeah I can open an issue for that too, I was able to just change the view angles for now.

If it's as easy as adding `focalpoint=None` to the signature and passing it through to `set_3d_view` under the hood somewhere, then it can be part of this PR

**Comment by alexrockhill** [comment]
Interestingly, it ran locally for me with the insula label but that was for `fsaverage` and `test_brain` uses `sample`. I don't think it's an issue with the core functionality though.

Actually that's for `aparc+aseg` and `fsaverage` and the test is for `aseg` and `sample`.

**Comment by larsoner** [comment]
> Interestingly, it ran locally for me with the insula label but that was for fsaverage and test_brain uses sample. I don't think it's an issue with the core functionality though.

I suspect it's a problem with the example arguments/usage. Can you fix it?

**Comment by alexrockhill** [comment]
> > Interestingly, it ran locally for me with the insula label but that was for fsaverage and test_brain uses sample. I don't think it's an issue with the core functionality though.
> 
> I suspect it's a problem with the example arguments/usage. Can you fix it?

Sure in a meeting, will fix in 30 minutes

**Comment by larsoner** [comment]
> Sure in a meeting, will fix in 30 minutes

Sounds good, FYI two examples appear broken now:

https://app.circleci.com/pipelines/github/mne-tools/mne-python/9151/workflows/e75f8447-9f58-40e4-b333-38bd240c79b9/jobs/31313

```
  File "/home/circleci/project/tutorials/inverse/60_visualize_stc.py", line 163, in <module>
    brain.add_volume_labels(aseg='aparc.a2009s+aseg', labels=labels)
  File "/home/circleci/project/mne/viz/_brain/_brain.py", line 2391, in add_volume_labels
    warn(f'Value not found for label {repr(label)} in: '
  File "/home/circleci/project/mne/utils/_logging.py", line 358, in warn
    warnings.warn_explicit(
RuntimeWarning: Value not found for label 'ctx_lh_G_occipital_sup' in: /home/circleci/mne_data/MNE-sample-data/subjects/sample/mri/aparc.a2009s+aseg.mgz
...
  File "/home/circleci/project/tutorials/clinical/20_seeg.py", line 178, in <module>
    brain.add_volume_labels(aseg='aparc+aseg', labels=labels)
  File "/home/circleci/project/mne/viz/_brain/_brain.py", line 2391, in add_volume_labels
    warn(f'Value not found for label {repr(label)} in: '
  File "/home/circleci/project/mne/utils/_logging.py", line 358, in warn
    warnings.warn_explicit(
RuntimeWarning: Value not found for label 'ctx-lh-caudalmiddlefrontal' in: /home/circleci/mne_data/MNE-sample-data/subjects/fsaverage/mri/aparc+aseg.mgz
```

**Comment by alexrockhill** [comment]
I think some changes maybe to smoothing have broken it, it's not finding any labels not just insula. Looking into it

**Comment by larsoner** [comment]
> I think some changes maybe to smoothing have broken i

The failure is in the `add_volume_labels` step, which should have nothing to do with labels, just with the contents of `aparc+aseg.mgz` and the FreeSurfer LUT. Or am I missing something?

**Comment by alexrockhill** [comment]
> > I think some changes maybe to smoothing have broken i
> 
> The failure is in the `add_volume_labels` step, which should have nothing to do with ~labels~ smoothing, just with the contents of `aparc+aseg.mgz` and the FreeSurfer LUT. Or am I missing something?

I'm going to try putting things back before your latest marching cubes change (only the first smooth), if that fixes things then we can do the discrete part in another PR.

**Comment by larsoner** [comment]
> I'm going to try putting things back before your latest marching cubes change (only the first smooth), if that fixes things then we can do the discrete part in another PR.

Feel free, but please preserve the warning. I suspect things might have been failing silently before (because we just silently didn't bother trying to do marching cubes before my commits) and the warning makes it clear that something is wrong somewhere (we shouldn't in an example ask to plot a volume label that does not exist in the mgz)

File: mne/viz/_brain/_brain.py
Comment (larsoner) [inline mne/viz/_brain/_brain.py]: This is the warning we need to preserve

File: mne/viz/_brain/_brain.py
Comment (larsoner) [inline mne/viz/_brain/_brain.py:2390]: ... you can add it back here

**Comment by alexrockhill** [comment]
> > I'm going to try putting things back before your latest marching cubes change (only the first smooth), if that fixes things then we can do the discrete part in another PR.
> 
> Feel free, but please preserve the warning. I suspect things might have been failing silently before (because we just silently didn't bother trying to do marching cubes before my commits) and the warning makes it clear that something is wrong somewhere (we shouldn't in an example ask to plot a volume label that does not exist in the mgz)

Good call, it does look like what was broken was from the marching cubes changes though. I didn't really delve into it fully but when I reverted all the surfaces rendered properly

**Comment by alexrockhill** [comment]
So there was an issue with `ctx-insula` not being found and that was because the atlas was changed to `aseg` which has a subset of labels, not including the `ctx` ones

**Comment by larsoner** [comment]
> Good call, it does look like what was broken was from the marching cubes changes though. I didn't really delve into it fully but when I reverted all the surfaces rendered properly

Did you verify this by locally testing with some `PATTERN=... make html_dev-pattern-memory` as suggested above (or even just running `python -i tutorials/clinical/...` to see if there is a warning)? You should get in the habit of doing this (example testing) as well as checking pytest -- a991d32 for example almost certainly broke `pytest mne/tests/test_surface.py -k march` because it reverted the array-like handling but not the tests I pushed for it. Try to get in the habit of testing these things locally before each push because:

1. It will make PR cycles shorter, and
2. We do end up with a backlog of CI runs in general

So pushing commits that uniformly keep CIs green is better than pushing commits without checking locally that things work first. The CIs are really meant to act as a sanity / completeness check of things that have been checked locally.

> So there was an issue with ctx-insula not being found and that was because the atlas was changed to aseg which has a subset of labels, not including the ctx ones

Can the discrete support be restored, or is there also an issue with the smoothing there?

**Comment by alexrockhill** [comment]
Yeah sorry, got a bit hasty with the fix for the last PR being a priority.

I checked and running locally there are no warnings for the `seeg` tutorial with this version, all the CIs should come back green.

> Can the discrete support be restored, or is there also an issue with the smoothing there?

There was an issue with the atlas in the tests but also a larger issue with the discrete method, I don't think it was working...

**Comment by larsoner** [comment]
> There was an issue with the atlas in the tests but also a larger issue with the discrete method, I don't think it was working...

Indeed it failed to find some ROIs, if I force it to use VTK_DOUBLE it seems to work fine! But like you said I can open a separate PR for this once this one is merged.

**Comment by larsoner** [comment]
Not waiting for macOS on this one, thanks @alexrockhill !

**Comment by alexrockhill** [comment]
Sounds great, looks like it's good to go

**Pull request merged**
